external/fv3fit/fv3fit/emulation/layers/architecture.py

```diff
@@ -178,3 +261,193 @@ def NoWeightSharingSLP(
             tf.keras.layers.Flatten(),
         ]
     )
+
+
+class StandardOutput(tf.keras.layers.Layer):
+    """Uses densely-connected layers w/ linear activation"""
+
+    def __init__(self, feature_lengths: Mapping[str, int], *args, **kwargs):
+        """
+        Args:
+            feature_lengths: Map of output variable names to expected
+                feature dimension length
+        """
+        super().__init__(*args, **kwargs)
+        self._feature_lengths = feature_lengths
+
+        self.output_layers = {
+            name: tf.keras.layers.Dense(feature_length, name=f"standard_output_{name}")
+            for name, feature_length in self._feature_lengths.items()
+        }
+
+    def call(self, hidden_output: tf.Tensor) -> Mapping[str, tf.Tensor]:
+        """
+        Args:
+            hidden_output: Network output from hidden layers
+
+        Returns:
+            Connected model outputs by variable
+        """
+
+        return {
+            name: layer(hidden_output) for name, layer in self.output_layers.items()
+        }
+
+
+def _single_dense(rnn_out, dense_layer):
+    """Passes a single dense layer across each vertical layer"""
+
+    nlevs = rnn_out.shape[-2]
+
+    connected = []
+    for i in range(nlevs):
+        connected.append(dense_layer(rnn_out[..., i, :]))
+
+    return tf.concat(connected, -2)
+
+
+class RNNOutput(tf.keras.layers.Layer):
+    """
+    Uses locally-connected layers w/ linear activation to
+    retain vertical dependence.
+
+    The output feature dimension is determined by the RNN output recursion dimension
+    length.  So to produce full output, the model inputs must contain the full
+    vertical dimension.
+
+    E.g., Combined inputs [nsamples, 17, 4] -> RNN network out [nsamples, 17, channels]
+        -> connected output [nsamples, 17]
+
+    If an output is a single-level (i.e., feature length of 1)
+    then it is connected to the lowest layer containing the full vertical information.
+    """
+
+    def __init__(
+        self,
+        feature_lengths: Mapping[str, int],
+        share_conv_weights=False,
+        *args,
+        **kwargs,
+    ):
+        """
+        Args:
+            feature_lengths: Map of output variable names to expected
+                feature dimension length
+        """
+        super().__init__(*args, **kwargs)
+        self._feature_lengths = feature_lengths
+        self._single_dense = share_conv_weights
+
+        if share_conv_weights:
+            layer_cls = tf.keras.layers.Conv1D
```

Comment:
I believe this would work for the "single dense model" suggestion.  Convolution shares weights for each vertical level instead of individual weights for each level.   Added to `ArchitectureConfig` under the name `rnn-v1-shared-weights`.